File: 11_Modular_TriggerBased_Puzzle_System/Unreal Scripts_&_Blueprints/PuzzleController.cpp
#include "PuzzleController.h"
#include "PuzzleTrigger.h"
#include "TimerManager.h"
#include "Engine/World.h"

APuzzleController::APuzzleController()
{
    PrimaryActorTick.bCanEverTick = false;
    bPuzzleComplete = false;
    DelayBeforeCheck = 0.0f;
}

void APuzzleController::BeginPlay()
{
    Super::BeginPlay();
}

void APuzzleController::AddPuzzleTrigger(APuzzleTrigger* Trigger)
{
    if (!Trigger) return;
    if (!PuzzleTriggers.Contains(Trigger))
    {
        PuzzleTriggers.Add(Trigger);
    }
}

void APuzzleController::CheckPuzzleState()
{
    if (!GetWorld()) return;

    bool bAllActivated = true;
    for (APuzzleTrigger* T : PuzzleTriggers)
    {
        if (!T || !T->IsActivated())
        {
            bAllActivated = false;
            break;
        }
    }

    if (bAllActivated && !bPuzzleComplete)
    {
        GetWorld()->GetTimerManager().ClearTimer(CheckTimerHandle);
        if (DelayBeforeCheck > 0.0f)
        {
            GetWorld()->GetTimerManager().SetTimer(CheckTimerHandle, this, &APuzzleController::DelayedCheck, DelayBeforeCheck, false);
        }
        else
        {
            CheckPuzzleStateMethod();
        }
    }
    else if (!bAllActivated && bPuzzleComplete)
    {
        GetWorld()->GetTimerManager().ClearTimer(CheckTimerHandle);
        CheckPuzzleStateMethod();
    }
}

void APuzzleController::DelayedCheck()
{
    CheckPuzzleStateMethod();
}

void APuzzleController::CheckPuzzleStateMethod()
{
    bool bAllActivated = true;

    for (APuzzleTrigger* T : PuzzleTriggers)
    {
        if (!T || !T->IsActivated())
        {
            bAllActivated = false;
            break;
        }
    }

    if (bAllActivated && !bPuzzleComplete)
    {
        bPuzzleComplete = true;
        OnPuzzleComplete.Broadcast();
    }
    else if (!bAllActivated && bPuzzleComplete)
    {
        bPuzzleComplete = false;
        OnPuzzleCancelComplete.Broadcast();
    }
}

File: 11_Modular_TriggerBased_Puzzle_System/Unity/PuzzleController.cs
using System.Collections;
using System.Collections.Generic;
using Unity.VisualScripting;
using UnityEngine;
using UnityEngine.Events;

public class PuzzleController : MonoBehaviour
{
    private List<PuzzleTrigger> puzzleTriggers = new List<PuzzleTrigger>();

    [Header("Puzzle State Events")]
    [Space(5)]
    public UnityEvent onPuzzleComplete;
    public UnityEvent onPuzzleCancelComplete;

    [Header("Optional Delay Before Checking Puzzle State")]
    [SerializeField] private float delayBeforeCheck = 0f;

    private bool puzzleComplete = false;

    public void AddPuzzleTrigger(PuzzleTrigger trigger)
    {
        if (!puzzleTriggers.Contains(trigger))
        {
            puzzleTriggers.Add(trigger);
        }
    }

    public void CheckPuzzleState()
    {
        if(delayBeforeCheck > 0f)
        {
            StartCoroutine(DelayedCheckPuzzleState());
            return;
        }
        else
        {
            CheckPuzzleStateMethod();
        }
    }

    private IEnumerator DelayedCheckPuzzleState()
    {
        yield return new WaitForSeconds(delayBeforeCheck);
        CheckPuzzleStateMethod();
    }

    public void CheckPuzzleStateMethod()
    {
        bool allActivated = true;

        foreach (var t in puzzleTriggers)
        {
            if (!t.isActivated)
            {
                allActivated = false;
                break;
            }
        }

        if (allActivated && !puzzleComplete)
        {
            puzzleComplete = true;
            onPuzzleComplete.Invoke();
        }
        else if (!allActivated && puzzleComplete)
        {
            puzzleComplete = false;
            onPuzzleCancelComplete.Invoke();
        }
    }
    
}

